tests/queries/0_stateless/03231_dynamic_variant_in_order_by_group_by.sql

```diff
@@ -0,0 +1,154 @@
+create table test (d Dynamic) engine=MergeTree order by map('str', d); -- {serverError DATA_TYPE_CANNOT_BE_USED_IN_KEY}
+create table test (d Dynamic) engine=MergeTree order by tuple() primary key d; -- {serverError DATA_TYPE_CANNOT_BE_USED_IN_KEY}
+create table test (d Dynamic) engine=MergeTree order by tuple() partition by d; -- {serverError DATA_TYPE_CANNOT_BE_USED_IN_KEY}
+create table test (d Dynamic) engine=MergeTree order by tuple() partition by tuple(d); -- {serverError DATA_TYPE_CANNOT_BE_USED_IN_KEY}
+create table test (d Dynamic) engine=MergeTree order by tuple() partition by array(d); -- {serverError DATA_TYPE_CANNOT_BE_USED_IN_KEY}
+create table test (d Dynamic) engine=MergeTree order by tuple() partition by map('str', d); -- {serverError DATA_TYPE_CANNOT_BE_USED_IN_KEY}
+
+create table test (d Variant(UInt64)) engine=MergeTree order by d; -- {serverError DATA_TYPE_CANNOT_BE_USED_IN_KEY}
+create table test (d Variant(UInt64)) engine=MergeTree order by tuple(d); -- {serverError DATA_TYPE_CANNOT_BE_USED_IN_KEY}
+create table test (d Variant(UInt64)) engine=MergeTree order by array(d); -- {serverError DATA_TYPE_CANNOT_BE_USED_IN_KEY}
+create table test (d Variant(UInt64)) engine=MergeTree order by map('str', d); -- {serverError DATA_TYPE_CANNOT_BE_USED_IN_KEY}
+create table test (d Variant(UInt64)) engine=MergeTree order by tuple() primary key d; -- {serverError DATA_TYPE_CANNOT_BE_USED_IN_KEY}
+create table test (d Variant(UInt64)) engine=MergeTree order by tuple() partition by d; -- {serverError DATA_TYPE_CANNOT_BE_USED_IN_KEY}
+create table test (d Variant(UInt64)) engine=MergeTree order by tuple() partition by tuple(d); -- {serverError DATA_TYPE_CANNOT_BE_USED_IN_KEY}
+create table test (d Variant(UInt64)) engine=MergeTree order by tuple() partition by array(d); -- {serverError DATA_TYPE_CANNOT_BE_USED_IN_KEY}
+create table test (d Variant(UInt64)) engine=MergeTree order by tuple() partition by map('str', d); -- {serverError DATA_TYPE_CANNOT_BE_USED_IN_KEY}
+
+create table test (d Dynamic) engine=Memory;
+insert into test select * from numbers(5);
+
+set allow_experimental_analyzer=1;
+
+set allow_suspicious_types_in_group_by=0;
+set allow_suspicious_types_in_order_by=0;
+
+select * from test order by d; -- {serverError ILLEGAL_COLUMN}
+select * from test order by tuple(d); -- {serverError ILLEGAL_COLUMN}
+select * from test order by array(d); -- {serverError ILLEGAL_COLUMN}
+select * from test order by map('str', d); -- {serverError ILLEGAL_COLUMN}
+
+select * from test group by d; -- {serverError ILLEGAL_COLUMN}
+select * from test group by tuple(d); -- {serverError ILLEGAL_COLUMN}
+select array(d) from test group by array(d); -- {serverError ILLEGAL_COLUMN}
+select map('str', d) from test group by map('str', d); -- {serverError ILLEGAL_COLUMN}
+select * from test group by grouping sets ((d), ('str')); -- {serverError ILLEGAL_COLUMN}
```

Comment:
With this small changes we can be sure that the two settings are not mixed somewhere in the codebase. As they are quite similar, have the same length etc, I think this might be a good thing to do.

```suggestion
set allow_suspicious_types_in_group_by=1;
set allow_suspicious_types_in_order_by=0;

select * from test order by d; -- {serverError ILLEGAL_COLUMN}
select * from test order by tuple(d); -- {serverError ILLEGAL_COLUMN}
select * from test order by array(d); -- {serverError ILLEGAL_COLUMN}
select * from test order by map('str', d); -- {serverError ILLEGAL_COLUMN}

set allow_suspicious_types_in_group_by=0;
set allow_suspicious_types_in_order_by=1;

select * from test group by d; -- {serverError ILLEGAL_COLUMN}
select * from test group by tuple(d); -- {serverError ILLEGAL_COLUMN}
select array(d) from test group by array(d); -- {serverError ILLEGAL_COLUMN}
select map('str', d) from test group by map('str', d); -- {serverError ILLEGAL_COLUMN}
select * from test group by grouping sets ((d), ('str')); -- {serverError ILLEGAL_COLUMN}
```

Reply:
If you think it makes sense, maybe it make sense to apply the same pattern in the rest of this file too.

Reply:
Good idea, updated test